File: src/main/java/seremis/geninfusion/api/soul/IGeneRegistry.java
package seremis.geninfusion.api.soul;

import net.minecraft.entity.EntityLiving;
import net.minecraft.item.ItemStack;
import seremis.geninfusion.api.soul.util.ModelPart;

import java.util.List;

public interface IGeneRegistry {

    IGene registerGene(String name, IGene gene);
    
    IGene registerGene(String name, EnumAlleleType type);

    IMasterGene registerMasterGene(String name, IMasterGene gene);

    void registerCustomInheritance(String name);

    void registerCustomInheritance(IGene gene);

    boolean useNormalInheritance(IGene gene);

    boolean useNormalInheritance(String name);

    List<IGene> getCustomInheritanceGenes();

    IGene getGene(String name);

    String getGeneName(IGene gene);

    int getGeneId(String name);

    int getGeneId(IGene gene);

    List<IGene> getGenes();

    IGene getGene(int id);

    ISoul getSoulFor(EntityLiving entity);

    IChromosome getChromosomeFor(EntityLiving entity, String name);

    IChromosome getChromosomeFor(IEntitySoulCustom entity, String name);

    IAllele getActiveFor(EntityLiving entity, String name);

    IAllele getActiveFor(IEntitySoulCustom entity, String name);

    boolean getValueBoolean(IEntitySoulCustom entity, String name);

    int getValueInteger(IEntitySoulCustom entity, String name);

    float getValueFloat(IEntitySoulCustom entity, String name);

    double getValueDouble(IEntitySoulCustom entity, String name);

    String getValueString(IEntitySoulCustom entity, String name);

    ItemStack getValueItemStack(IEntitySoulCustom entity, String name);

    ModelPart getValueModelPart(IEntitySoulCustom entity, String name);

    Class getValueClass(IEntitySoulCustom entity, String name);

    boolean[] getValueBooleanArray(IEntitySoulCustom entity, String name);

    int[] getValueIntegerArray(IEntitySoulCustom entity, String name);

    float[] getValueFloatArray(IEntitySoulCustom entity, String name);

    double[] getValueDoubleArray(IEntitySoulCustom entity, String name);

    String[] getValueStringArray(IEntitySoulCustom entity, String name);

    ItemStack[] getValueItemStackArray(IEntitySoulCustom entity, String name);

    ModelPart[] getValueModelPartArray(IEntitySoulCustom entity, String name);

    Class[] getValueClassArray(IEntitySoulCustom entity, String name);

    List<String> getControlledGenes(String masterGeneName);
}

File: src/main/scala/seremis/geninfusion/api/soul/IGeneRegistry.scala
package seremis.geninfusion.api.soul

import net.minecraft.entity.EntityLiving

trait IGeneRegistry {

    def registerGene(name: String, gene: IGene): IGene

    def registerGene(name: String, clzz: Class[_]): IGene

    def registerMasterGene(name: String, gene: IMasterGene): IMasterGene

    def registerCustomInheritance(name: String)

    def registerCustomInheritance(gene: IGene)

    def useNormalInheritance(gene: IGene): Boolean

    def useNormalInheritance(name: String): Boolean

    def getCustomInheritanceGenes: List[IGene]

    def getGene(name: String): Option[IGene]

    def getGene(id: Int): Option[IGene]

    def getGeneName(gene: IGene): Option[String]

    def getGeneId(name: String): Option[Int]

    def getGeneId(gene: IGene): Option[Int]

    def getGenes: List[IGene]

    def getSoulFor(entity: EntityLiving): Option[ISoul]

    def getChromosomeFor(entity: EntityLiving, name: String): Option[IChromosome]

    def getChromosomeFor(entity: IEntitySoulCustom, name: String): Option[IChromosome]

    def getActiveFor(entity: EntityLiving, name: String): Option[IAllele]

    def getActiveFor(entity: IEntitySoulCustom, name: String): Option[IAllele]

    def getValueFromAllele[T](entity: IEntitySoulCustom, name: String): T

    /**
     * Change the value of an allele of the gene with the specified name. This is only possible for genes with the flag isChangeable true. (see IGene)
     * @param entity The entity for which to change the allele.
     * @param name The name of the IGene for which to change the allele.
     * @param value The new value of the specified IGene.
     * @param changeActiveGene Whether to change the active or recessive gene.
     */
    def changeAlleleValue[T](entity: IEntitySoulCustom, name: String, value: T, changeActiveGene: Boolean)

    def getControlledGenes(masterGeneName: String): List[String]
}
